### PR TITLE
Breaking: Move MXNet-based models to `mx.model`.

### LINKED ISSUE #2117: Move MXNet based models from `gluonts.model` to `gluonts.mx`.
When we started, we put all models into `gluonts.model`. However, when adding support for PyTorch, we started to put these models into `gluonts.torch.model`, since they were re-implementations of existing models.

Ideally, we replicate this setup for `mxnet` based models, and move all these models to `gluonts.mx.model`.

Further we can then import models directly from `mx`:

```py
from gluonts.mx import DeepAREstimator
```

For `v0.11` we add a deprecation warning and remove the warning in `v0.12.`

### PR DESCRIPTION
**Pull request by Hongqing-work**
Closes: #2117

*Description of changes:*
move origin models from `gluonts.model` to `gluonts.mx.model` and use `from gluonts.mx import xxEstimator`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by jaheba** [comment]
Thanks!

Something we want to make sure is to still re-export the models using their old locations, but emit a deprecation warning if one uses the old path.

That way we can more gracefully make the transition.

File: README.md [hunk 1]
```diff
@@ -36,7 +36,7 @@ of data.
 from gluonts.dataset.util import to_pandas
 from gluonts.dataset.pandas import PandasDataset
 from gluonts.dataset.repository.datasets import get_dataset
-from gluonts.model.deepar import DeepAREstimator
+from gluonts.mx import DeepAREstimator
```
Comment (jaheba) [inline README.md]: Let's maybe not change this in the README for now.

Users expect the example from the GitHub project page to work with the current version of GluonTS.

File: docs/tutorials/advanced_topics/trainer_callbacks.md.input [hunk 2]
```diff
@@ -18,7 +18,7 @@ To use callbacks, simply pass them as a list when constructing the `Trainer`:
 in the following example, we are using the `TrainingHistory` callback to record loss values measured during training.
 
 ```python
-from gluonts.model.simple_feedforward import SimpleFeedForwardEstimator
+from gluonts.mx import SimpleFeedForwardEstimator
 from gluonts.mx import Trainer
```
Comment (jaheba) [inline docs/tutorials/advanced_topics/trainer_callbacks.md.input]: Can we merge these into one line?

File: docs/tutorials/advanced_topics/howto_pytorch_lightning.md.input [hunk 3]
```diff
@@ -57,7 +57,7 @@ import torch.nn as nn
 ```python
 from gluonts.torch.model.predictor import PyTorchPredictor
 from gluonts.torch.distributions import StudentTOutput
-from gluonts.model.forecast_generator import DistributionForecastGenerator
+from gluonts.mx.forecast_generator import DistributionForecastGenerator
```
Comment (jaheba) [inline docs/tutorials/advanced_topics/howto_pytorch_lightning.md.input]: AFAIK `forecast_generator` is generic and should not move to `mx`.

File: docs/tutorials/data_manipulation/pandasdataframes.md.input [hunk 4]
```diff
@@ -83,7 +83,7 @@ ds = PandasDataset.from_long_dataframe(df, target="target", item_id="item_id")
 ```
 
 ```python
-from gluonts.model.deepar import DeepAREstimator
+from gluonts.mx import DeepAREstimator
 from gluonts.mx import Trainer
```
Comment (jaheba) [inline docs/tutorials/data_manipulation/pandasdataframes.md.input]: Can we merge these into one line?

File: docs/tutorials/data_manipulation/pandasdataframes.md.input [hunk 5]
```diff
@@ -227,7 +227,7 @@ to fully train the model.
 
 
 ```python
-from gluonts.model.deepar import DeepAREstimator
+from gluonts.mx import DeepAREstimator
 from gluonts.mx import Trainer
```
Comment (jaheba) [inline docs/tutorials/data_manipulation/pandasdataframes.md.input]: Can we merge these into one line?

File: docs/tutorials/forecasting/extended_tutorial.md.input [hunk 6]
```diff
@@ -573,7 +573,7 @@ Finally, each estimator is configured by a `Trainer`, which defines how the mode
 
 
 ```python
-from gluonts.model.simple_feedforward import SimpleFeedForwardEstimator
+from gluonts.mx import SimpleFeedForwardEstimator
 from gluonts.mx import Trainer
```
Comment (jaheba) [inline docs/tutorials/forecasting/extended_tutorial.md.input]: Can we merge these into one line?

File: docs/tutorials/forecasting/quick_start_tutorial.md.input [hunk 7]
```diff
@@ -130,7 +130,7 @@ Finally, each estimator is configured by a `Trainer`, which defines how the mode
 
 
 ```python
-from gluonts.model.simple_feedforward import SimpleFeedForwardEstimator
+from gluonts.mx import SimpleFeedForwardEstimator
 from gluonts.mx import Trainer
```
Comment (jaheba) [inline docs/tutorials/forecasting/quick_start_tutorial.md.input]: Can we merge these into one line?

File: src/gluonts/model/gp_forecaster/__init__.py [hunk 8]
```diff
@@ -11,6 +11,15 @@
 # express or implied. See the License for the specific language governing
 # permissions and limitations under the License.
 
-from ._estimator import GaussianProcessEstimator
+import warnings
+
+from gluonts.mx.model.gp_forecaster._estimator import GaussianProcessEstimator
```
Comment (jaheba) [inline src/gluonts/model/gp_forecaster/__init__.py]: I think in all these, it's not necessary to import from `_estimator`:

```suggestion
from gluonts.mx.model.gp_forecaster import GaussianProcessEstimator
```

File: src/gluonts/model/trivial/constant.py [hunk 9]
```diff
@@ -18,7 +18,10 @@
 from gluonts.dataset.field_names import FieldName
 from gluonts.dataset.util import forecast_start
 from gluonts.model.forecast import SampleForecast
-from gluonts.model.predictor import FallbackPredictor, RepresentablePredictor
+from gluonts.model.predictor import (
+    FallbackPredictor,
+    RepresentablePredictor,
+)
```
Comment (jaheba) [inline src/gluonts/model/trivial/constant.py]: Can we undo the changes?

The trailing `,` means that black will format these on multiple lines. If you remove it, black should format it on one line.

File: src/gluonts/model/trivial/mean.py [hunk 10]
```diff
@@ -22,7 +22,10 @@
 from gluonts.dataset.util import forecast_start
 from gluonts.model.estimator import Estimator
 from gluonts.model.forecast import SampleForecast
-from gluonts.model.predictor import FallbackPredictor, RepresentablePredictor
+from gluonts.model.predictor import (
+    FallbackPredictor,
+    RepresentablePredictor,
+)
```
Comment (jaheba) [inline src/gluonts/model/trivial/mean.py]: Sam as above.

File: src/gluonts/model/wavenet/__init__.py [hunk 11]
```diff
@@ -11,7 +11,16 @@
 # express or implied. See the License for the specific language governing
 # permissions and limitations under the License.
 
-from ._estimator import WaveNetEstimator
-from ._network import WaveNet, WaveNetSampler
+import warnings
+
+from gluonts.mx.model.wavenet._estimator import WaveNetEstimator
+from gluonts.mx.model.wavenet._network import WaveNet, WaveNetSampler
```
Comment (jaheba) [inline src/gluonts/model/wavenet/__init__.py]: Again, no need to import from `_estimator` and `_network.`

File: src/gluonts/mx/model/deepstate/_estimator.py [hunk 12]
```diff
@@ -27,7 +27,7 @@
     ValidationDataLoader,
 )
 from gluonts.env import env
-from gluonts.model.deepstate.issm import ISSM, CompositeISSM
+from gluonts.mx.model.deepstate.issm import ISSM, CompositeISSM
```
Comment (jaheba) [inline src/gluonts/mx/model/deepstate/_estimator.py]: This should be a relative import

```suggestion
from .issm import ISSM, CompositeISSM
```

Then we don't have to change it whenever we move this around.

File: src/gluonts/nursery/autogluon_tabular/predictor.py [hunk 13]
```diff
@@ -23,7 +23,11 @@
 from gluonts.dataset.util import to_pandas
 from gluonts.dataset.field_names import FieldName
 from gluonts.itertools import batcher
-from gluonts.model.forecast import SampleForecast, QuantileForecast, Forecast
+from gluonts.model.forecast import (
+    SampleForecast,
+    QuantileForecast,
+    Forecast,
+)
```
Comment (jaheba) [inline src/gluonts/nursery/autogluon_tabular/predictor.py]: Can we also revert this?

File: test/model/transformer/_estimator.py [hunk 14]
```diff
@@ -0,0 +1,389 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (jaheba) [inline test/model/transformer/_estimator.py]: This file looks like it was accidentally added?

File: src/gluonts/model/canonical/__init__.py [hunk 15]
```diff
@@ -11,6 +11,16 @@
...
+import warnings
+
+from gluonts.mx.model.canonical import CanonicalRNNEstimator
+
+warnings.warn(
+    "The module gluonts.model.canonical has been moved to "
+    "gluonts.mx.model.canonical. In GluonTS v0.12 it will be no longer "
+    "possible to use the old path. Try to use 'from gluonts.mx import "
+    "CanonicalRNNEstimator'.",
+    FutureWarning,
```
Comment (jaheba) [inline src/gluonts/model/canonical/__init__.py:23]: I think there is still something confusing about this.

We should think about what the canonical path of the models should be. In this case is `gluonts.mx.model.canonical` or `gluonts.mx` the thing that should be remembered?

I like the idea of exposing the models directly in `mx` and ideally people don't need to know that the `model` sub-package exists.

@lostella thoughts?

File: src/gluonts/mx/__init__.py [hunk 16]
```diff
@@ -37,3 +37,38 @@
 )
 from .trainer import Trainer
 from .util import copy_parameters, get_hybrid_forward_input_names
+
+from .model import forecast_generator
```
Comment (jaheba) [inline src/gluonts/mx/__init__.py]: Shouldn't this use the `gluonts.model` package?

File: src/gluonts/mx/model/deepstate/_estimator.py [hunk 17]
```diff
@@ -27,7 +27,7 @@
     ValidationDataLoader,
 )
 from gluonts.env import env
-from gluonts.model.deepstate.issm import ISSM, CompositeISSM
+from .issm import ISSM, CompositeISSM
```
Comment (jaheba) [inline src/gluonts/mx/model/deepstate/_estimator.py]: Can you also move these imports to the end of the block?

File: src/gluonts/mx/__init__.py [hunk 18]
```diff
@@ -37,3 +37,36 @@
 )
 from .trainer import Trainer
 from .util import copy_parameters, get_hybrid_forward_input_names
+
```
Comment (jaheba) [inline src/gluonts/mx/__init__.py:40]: Can we add all the imports to `__all__`?

**Review by jaheba — APPROVED** [review]
